[DOCS-14269] Add private setup page for OSS OTel Collector (#36452)
DataDog/documentation: content/en/opentelemetry/setup/collector_exporter/oss_setup.md

title: Set Up the OpenTelemetry Collector (OSS)
private: true
description: 'Send OpenTelemetry data to Datadog using the OpenTelemetry Collector with standard OSS components'
further_reading:
- link: "https://opentelemetry.io/docs/collector/"
  tag: "External Site"
  text: "Collector documentation"
- link: "/opentelemetry/setup/collector_exporter/deploy"
  tag: "Documentation"
  text: "Deploy the OpenTelemetry Collector"
- link: "/opentelemetry/config/hostname_tagging"
  tag: "Documentation"
  text: "Configure Hostname and Tagging"
---

## Overview

Send traces, metrics, and logs to Datadog using the [OpenTelemetry Collector Contrib][1] distribution with standard OpenTelemetry components. This setup uses the following key components:

- **OTLP HTTP exporter**: Sends telemetry to Datadog's OTLP intake endpoints.
- **Span metrics connector**: Generates RED (Rate, Error, Duration) metrics from trace data to power APM features like the Service Catalog and Service Page.
- **Resource detection processor**: Extracts host and cloud metadata for hostname resolution and tagging in Datadog.

<!-- TODO: Replace with updated diagram showing OTLP HTTP exporter flow (no Datadog Exporter) -->

> <div class="alert alert-info">To see which Datadog features are supported with this setup, see the <a href="/opentelemetry/compatibility/">feature compatibility table</a> under <b>OTel SDK + OSS Collector</b>.</div>

Reviewer: Nit: Maybe there should be a general warning about this setup being experimental and potentially having slight differences with the output of an equivalent DD exporter setup? Notably, the host list will have less information until we implement host metadata ingestion. But maybe that's a given for a preview.

## Prerequisites

- [OpenTelemetry Collector Contrib][1] v0.150.1 or later
- A [Datadog API key][2]
- Your [Datadog site][3] (for example, `datadoghq.com` or `datadoghq.eu`)

## Install and configure

### 1. Download the OpenTelemetry Collector

> Download the latest release of the OpenTelemetry Collector Contrib distribution from the [releases page][100].

Reviewer: Nit: This instruction is only relevant for customers in the Host-based deployment, so maybe it should be included under the tabs? Although [the existing Collector page](https://docs-staging.datadoghq.com/brett.blue/otel-collector-getting-started/opentelemetry/setup/collector_exporter/) does the same thing so this can be left to a later PR.

      - name: k8s.replicaset.name
      - name: k8s.statefulset.name
      - name: k8s.daemonset.name
      - name: k8s.job.name
      - name: k8s.cronjob.name
      - name: k8s.namespace.name
      - name: k8s.pod.name
      - name: aws.ecs.launchtype
      - name: cloud.account.id
      - name: host.id
      - name: host.name
      - name: k8s.node.name
      - name: azure.resourcegroup.name
      - name: aws.s3.bucket
      - name: db.namespace
      - name: messaging.destination.name
      - name: messaging.system
      - name: server.address
      - name: operation.name
      - name: http.request.method
      - name: http.method
      - name: db.system
      - name: messaging.operation
      - name: rpc.system
      - name: rpc.service
      - name: faas.invoked_provider
      - name: faas.invoked_name
      - name: faas.trigger
      - name: graphql.operation.type
      - name: network.protocol.name
      - name: resource.name
      - name: http.route
      - name: messaging.destination
      - name: rpc.method
      - name: graphql.operation.name
      - name: db.statement
      - name: db.query.text

exporters:
  otlp_http:
    endpoint: https://otlp.${env:DD_SITE}

>     metrics_endpoint: https://otlp.${env:DD_SITE}/api/v2/otlpmetrics

Reviewer: @dineshg13 Do you know if there are plans to enable the new metrics endpoint on the standard path for preview, or would that come later?

    headers:
      dd-api-key: ${env:DD_API_KEY}
      dd-otel-metric-config: >-
        {
        "resource_attributes_as_tags": true,
        "instrumentation_scope_metadata_as_tags": true
        }
    compression: zstd
    compression_params:
      level: 3
    sending_queue:
      batch: {}

extensions:
  datadog:
    api:
      site: ${env:DD_SITE}
      key: ${env:DD_API_KEY}
    deployment_type: daemonset

service:
  extensions:
    - datadog
  pipelines:
    logs:
      receivers: [otlp]
      processors: [resourcedetection]
      exporters: [otlp_http]
    metrics:
      receivers: [otlp, hostmetrics]
      processors: [resourcedetection, cumulativetodelta]
      exporters: [otlp_http]
    traces:
      receivers: [otlp]
      processors: [resourcedetection]
      exporters: [forward, spanmetrics]
    traces/sampling:
      receivers: [forward]

>       exporters: [otlp_http]

Reviewer: Nit: I think it's fine to remove some of the comments from the config, but I think some are somewhat load-bearing, like the one here about "this is where you would insert your sampling processors"; otherwise you may as well add the `otlp_http` exporter directly to `traces`

        "instrumentation_scope_metadata_as_tags": true
        }
    compression: zstd
    compression_params:
      level: 3
    sending_queue:
      batch: {}

extensions:
  health_check:
    endpoint: ${env:MY_POD_IP}:13133
  datadog:
    api:
      site: ${env:DD_SITE}
      key: ${env:DD_API_KEY}
    deployment_type: daemonset

service:
  extensions:
    - health_check
    - datadog
  pipelines:
    logs:
      receivers: [otlp]
      processors: [k8s_attributes, resourcedetection]
      exporters: [otlp_http]
    metrics:
      receivers: [otlp, hostmetrics]
      processors: [k8s_attributes, resourcedetection, cumulativetodelta]
      exporters: [otlp_http]
    traces:
      receivers: [otlp]
      processors: [k8s_attributes, resourcedetection]
      exporters: [forward, spanmetrics]
    traces/sampling:
      receivers: [forward]
      exporters: [otlp_http]
    metrics/spanmetrics:
      receivers: [spanmetrics]
      exporters: [otlp_http]
  telemetry:
    metrics:
      readers:
        - periodic:
            exporter:
              otlp:
                protocol: http/protobuf
                endpoint: http://localhost:4318
```

> For EKS environments, add the `eks` and `ec2` detectors to the `resourcedetection` processor:

Reviewer: Nit: I didn't generate a full config based on it yet, but similar cluster name config [is needed for AKS](https://github.com/DataDog/opentelemetry-examples/blob/c2b4e39985017e105125881d599c98f9ae696b27/configurations/opentelemetry-collector/generator/templates/components/resource-detection-processor.tmpl#L22-L24)

```yaml
processors:
  resourcedetection:
    detectors: [eks, ec2, env]
    timeout: 2s
    override: true
    eks:
      resource_attributes:
        k8s.cluster.name: { enabled: true }
    ec2:
      tags: ['^kubernetes\.io/cluster/.*$']
```

The `k8s_attributes` processor requires a ServiceAccount with permissions to read pod metadata. See the [Kubernetes Attributes Processor documentation][101] for RBAC setup instructions.

[101]: https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/k8sattributesprocessor#role-based-access-control

{{% /tab %}}

> {{% tab "Kubernetes (Helm)" %}}

Reviewer: Nit: This is a bit misleading since this is still a Daemonset deployment, just deployed a different way. (And the Collector Helm chart can do other types of deployments as well)

You can deploy the Collector in Kubernetes using the [official OpenTelemetry Collector Helm chart][102].

1. Create a Kubernetes secret with your Datadog API key:

   ```shell
   kubectl create secret generic datadog-secrets --from-literal=api-key='<YOUR_API_KEY>'
   ```

1. Add the OpenTelemetry Helm repository:

   ```shell
   helm repo add open-telemetry https://open-telemetry.github.io/opentelemetry-helm-charts
   ```

> 1. Download the [example values file][103] and save it as `values.yaml`. This file configures the Collector as a DaemonSet with the recommended Datadog settings.

Reviewer: There are many cases where customers will want to modify the `values.yaml`, but one almost required modification is checking the value of `DD_SITE` if their site isn't `datadoghq.com`, so I would call it out

1. Install the Collector:

   ```shell
   helm install otelcol open-telemetry/opentelemetry-collector --values values.yaml
   ```

[102]: https://github.com/open-telemetry/opentelemetry-helm-charts/tree/main/charts/opentelemetry-collector
[103]: https://github.com/DataDog/opentelemetry-examples/blob/experimental-oss-config/configurations/opentelemetry-collector/helm-daemonset.yaml

{{% /tab %}}
{{< /tabs >}}

### 3. Run the Collector

> Start the Collector with the required feature gate enabled. If you are using Docker or Kubernetes, the run command is included in the configuration tab above.

Reviewer 2: nit: For most users this is not required, but recommended, I think we could soften the language here
```suggestion
Start the Collector with the recommended feature gate enabled. If you are using Docker or Kubernetes, the run command is included in the configuration tab above.
```

For Host installations, run:

```shell
DD_SITE={{< region-param key="dd_site" >}} DD_API_KEY=<YOUR_API_KEY> \
  otelcol-contrib --config collector.yaml \
  --feature-gates connector.spanmetrics.includeCollectorInstanceID
```

### 4. Configure your application

Configure your OpenTelemetry-instrumented application to send data to the Collector. Set the `OTEL_EXPORTER_OTLP_ENDPOINT` environment variable to point to the Collector:

{{< tabs >}}
{{% tab "Host" %}}
```shell
export OTEL_EXPORTER_OTLP_ENDPOINT="http://localhost:4318"

> ```

Reviewer 2: nit: this is not needed (it's [the default value](https://opentelemetry.io/docs/languages/sdk-configuration/otlp-exporter/#otel_exporter_otlp_endpoint)). I don't know if getting rid of the tab is a good idea or if we just leave as is

Reviewer 3: I believe also the default protocol varies by language (some are `grpc` and some are `http/protobuf`) so this value won't work for all langauges? (I'm not sure what happens if you configure the http endpoint but the protocol is configured for `grpc`)

Reviewer 2: Right, I guess we should set [`OTEL_EXPORTER_OTLP_PROTOCOL`](https://opentelemetry.io/docs/languages/sdk-configuration/otlp-exporter/#otel_exporter_otlp_protocol) if we expect a fully env-var based configuration

After your application sends telemetry to the Collector, verify that data appears in Datadog:

1. In Datadog, go to **APM > Services** and confirm that your `service.name` appears.
2. Open **APM > Traces** and search for your service.
3. Go to **Infrastructure > Host Map** and confirm that the host running the Collector appears.
4. If you send logs through OTLP, go to **Logs Explorer** and search for your service name.

## Key components

### Span metrics connector

The `spanmetrics` connector generates RED metrics from trace data. These metrics power APM features including the Service Catalog, Service Page, and Resource Page. The connector is configured with dimensions that enable Datadog to compute host tags, peer services, and operation names from your traces.

For a complete list of dimensions included in the recommended configuration, see the [full configuration files][5] in the `opentelemetry-examples` repository.

### OTLP HTTP exporter

The `otlp_http` exporter sends telemetry data to Datadog's OTLP intake endpoints. Key configuration details:

- **Endpoint**: `https://otlp.{your_dd_site}` for traces and logs, `https://otlp.{your_dd_site}/api/v2/otlpmetrics` for metrics.

> - **Compression**: `zstd` is recommended for reduced bandwidth usage.

Reviewer: Nit: Maybe we should call out that `compression_params::level` is basically required, because the default for zstd is the lowest level of compression?

---
title: Set Up the OpenTelemetry Collector (OSS)
private: true
description: 'Send OpenTelemetry data to Datadog using the OpenTelemetry Collector with standard OSS components'
further_reading:
- link: "https://opentelemetry.io/docs/collector/"
  tag: "External Site"
  text: "Collector documentation"
- link: "/opentelemetry/setup/collector_exporter/deploy"
  tag: "Documentation"
  text: "Deploy the OpenTelemetry Collector"
- link: "/opentelemetry/config/hostname_tagging"
  tag: "Documentation"
  text: "Configure Hostname and Tagging"
---

## Overview

Send traces, metrics, and logs to Datadog using the [OpenTelemetry Collector Contrib][1] distribution with standard OpenTelemetry components. This setup uses the following key components:

- **OTLP HTTP exporter**: Sends telemetry to Datadog's OTLP intake endpoints.

> - **Span metrics connector**: Generates RED (Rate, Error, Duration) metrics from trace data to power APM features like the Service Catalog and Service Page.

Reviewer 4: ```suggestion
- **Span metrics connector**: Generates RED (Rate, Error, Duration) metrics from trace data to power APM features such as the Service Catalog and Service Page.
```

        metrics:
          system.memory.limit:
            enabled: true
      paging:
        metrics:
          system.paging.utilization:
            enabled: true
          system.paging.usage:
            enabled: true
      disk: {}
      filesystem:
        metrics:
          system.filesystem.utilization:
            enabled: true
      load: {}
      network: {}
      processes: {}

processors:
  # Detect host and cloud metadata for hostname resolution and tagging
  resourcedetection:
    detectors: [env, system]
    timeout: 2s
    override: true
    system:
      resource_attributes:
        host.arch:
          enabled: true
        host.cpu.cache.l2.size:
          enabled: true
        host.cpu.family:
          enabled: true
        host.cpu.model.id:
          enabled: true
        host.cpu.model.name:
          enabled: true
        host.cpu.stepping:
          enabled: true
        host.cpu.vendor.id:
          enabled: true
        host.ip:
          enabled: true
        host.mac:
          enabled: true
        os.description:
          enabled: true
  # Convert cumulative metrics to delta temporality for Datadog
  cumulativetodelta: {}

connectors:

>   # Separates trace processing from sampling so span metrics are computed on all traces

Reviewer 4: ```suggestion
  # Separate trace processing from sampling so span metrics are computed on all traces
```

Reviewer 4: Other comments in the same configs use imperative voice ("Receive...", "Collect...", "Detect...", "Convert...", "Generate...")

  forward: {}
  # Generate RED (Rate, Error, Duration) metrics from traces for APM
  spanmetrics:
    aggregation_temporality: AGGREGATION_TEMPORALITY_DELTA
    add_resource_attributes: true
    histogram:
      exponential: {}
      unit: s

>     dimensions:

Reviewer 5: @mx-psi Can we add a comment here explaining what this powers?

Reviewer 2: There are multiple uses for this section, the original config mentions them very succintly https://github.com/DataDog/opentelemetry-examples/blob/c2b4e39985017e105125881d599c98f9ae696b27/configurations/opentelemetry-collector/otel-demo-eks.yaml#L187-L267

Maybe we can recover those comments? We could include a link to the specific product feature they power if we want to be a bit more explicit

      http:
        endpoint: 0.0.0.0:4318
  # Collect host-level metrics for the Infrastructure List
  # root_path maps to the host filesystem mounted at /hostfs
  hostmetrics:
    root_path: /hostfs
    collection_interval: 10s
    scrapers:
      cpu:
        metrics:
          system.cpu.utilization:
            enabled: true
          system.cpu.physical.count:
            enabled: true
          system.cpu.logical.count:
            enabled: true
          system.cpu.frequency:
            enabled: true
      memory:
        metrics:
          system.memory.limit:
            enabled: true
      paging:
        metrics:
          system.paging.utilization:
            enabled: true
          system.paging.usage:
            enabled: true
      disk: {}
      filesystem:
        metrics:
          system.filesystem.utilization:
            enabled: true
      load: {}
      network: {}
      processes: {}

processors:
  # Detect host and cloud metadata for hostname resolution and tagging
  resourcedetection:
    detectors: [env]
    timeout: 2s
    override: true
  # Convert cumulative metrics to delta temporality for Datadog
  cumulativetodelta: {}

connectors:

>   # Separates trace processing from sampling so span metrics are computed on all traces

Reviewer 4: ```suggestion
  # Separate trace processing from sampling so span metrics are computed on all traces
```

  resourcedetection:
    detectors: [env]
    timeout: 2s
    override: true
  # Convert cumulative metrics to delta temporality for Datadog
  cumulativetodelta: {}
  # Enrich telemetry with Kubernetes pod and container metadata
  k8s_attributes:
    extract:
      otel_annotations: true
      metadata:
        - k8s.node.name
        - k8s.namespace.name
        - service.namespace
        - service.name
        - service.version
        - service.instance.id
        - k8s.deployment.name
        - k8s.replicaset.name
        - k8s.daemonset.name
        - k8s.statefulset.name
        - k8s.cronjob.name
        - k8s.job.name
        - k8s.pod.uid
        - k8s.pod.name
        - container.id
        - k8s.container.name
        - container.image.name
        - container.image.tag
    pod_association:
      - sources:
          - from: resource_attribute
            name: k8s.pod.uid
      - sources:
          - from: resource_attribute
            name: k8s.pod.ip
      - sources:
          - from: resource_attribute
            name: k8s.pod.name
          - from: resource_attribute
            name: k8s.namespace.name
      - sources:
          - from: connection

connectors:

>   # Separates trace processing from sampling so span metrics are computed on all traces

Reviewer 4: ```suggestion
  # Separate trace processing from sampling so span metrics are computed on all traces
```

[101]: https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/k8sattributesprocessor#role-based-access-control

{{% /tab %}}

{{% tab "Kubernetes (Helm chart)" %}}

You can deploy the Collector as a DaemonSet in Kubernetes using the [official OpenTelemetry Collector Helm chart][102].

1. Create a Kubernetes secret with your Datadog API key:

   ```shell
   kubectl create secret generic datadog-secrets --from-literal=api-key='<YOUR_API_KEY>'
   ```

1. Add the OpenTelemetry Helm repository:

   ```shell
   helm repo add open-telemetry https://open-telemetry.github.io/opentelemetry-helm-charts
   ```

1. Download the [example values file][103] and save it as `values.yaml`. This file configures the Collector as a DaemonSet with the recommended Datadog settings. If your Datadog site is not `datadoghq.com`, update the `DD_SITE` value in `values.yaml` before installing.

1. Install the Collector:

   ```shell
   helm install otelcol open-telemetry/opentelemetry-collector --values values.yaml
   ```

[102]: https://github.com/open-telemetry/opentelemetry-helm-charts/tree/main/charts/opentelemetry-collector

> [103]: https://github.com/DataDog/opentelemetry-examples/blob/experimental-oss-config/configurations/opentelemetry-collector/helm-daemonset.yaml

Reviewer 4: This link likewise seems to be pointing to a non-main branch, but I'm not finding this file in the main branch. Could you please confirm this is expected?

Author: Yep! This is expected for now. Thanks for the detailed look 👀

{{% /tab %}}
{{< /tabs >}}

### 3. Run the Collector

> Start the Collector with the recommended feature gate enabled. If you are using Docker or Kubernetes, the run command is included in the configuration tab above.

Reviewer 4: ```suggestion
Start the Collector with the recommended feature gate enabled. If you are using Docker or Kubernetes, the run command is included in the [Create the collector configuration](#2-create-the-collector-configuration) section.
```

## Verify the setup

After your application sends telemetry to the Collector, verify that data appears in Datadog:

1. In Datadog, go to **APM > Services** and confirm that your `service.name` appears.
2. Open **APM > Traces** and search for your service.
3. Go to **Infrastructure > Host Map** and confirm that the host running the Collector appears.
4. If you send logs through OTLP, go to **Logs Explorer** and search for your service name.

## Key components

### Span metrics connector

The `spanmetrics` connector generates RED metrics from trace data. These metrics power APM features including the Service Catalog, Service Page, and Resource Page. The connector is configured with dimensions that enable Datadog to compute host tags, peer services, and operation names from your traces.

For a complete list of dimensions included in the recommended configuration, see the [full configuration files][5] in the `opentelemetry-examples` repository.

### OTLP HTTP exporter

The `otlp_http` exporter sends telemetry data to Datadog's OTLP intake endpoints. Key configuration details:

> - **Endpoint**: `https://otlp.{your_dd_site}` for traces and logs, `https://otlp.{your_dd_site}/api/v2/otlpmetrics` for metrics.

Reviewer 4: ```suggestion
- **Endpoint**: `https://otlp.<YOUR_DD_SITE>` for traces and logs, `https://otlp.<YOUR_DD_SITE>/api/v2/otlpmetrics` for metrics.
```

- **Compression**: `zstd` is recommended for reduced bandwidth usage. When using `zstd`, set `compression_params.level` explicitly, because the default uses the lowest compression level.
- **Resource attributes as tags**: The `dd-otel-metric-config` header enables resource attributes and instrumentation scope metadata to be sent as metric tags.

### Datadog extension

> The `datadog` extension sends Collector metadata to Datadog for host enrichment. It does not export telemetry data—all telemetry flows through the OTLP HTTP exporter. This extension is part of the [OpenTelemetry Collector Contrib][1] project and handles API key validation and deployment type reporting.

Reviewer 4: ```suggestion
The `datadog` extension sends Collector metadata to Datadog for host enrichment. It does not export telemetry data. All telemetry flows through the OTLP HTTP exporter. This extension is part of the [OpenTelemetry Collector Contrib][1] project and handles API key validation and deployment type reporting.
```

Reviewer 4: Per the editorial reference on em-dashes:

> Don't use to set apart a clause that could stand as its own sentence; use a period instead.

### Cumulative-to-delta processor

The `cumulativetodelta` processor converts cumulative metrics to delta temporality, which is [Datadog's recommended configuration][6] for OpenTelemetry metrics.

### Self-monitoring telemetry

The configuration sends the Collector's own metrics back to its local OTLP receiver (`http://localhost:4318`). This routes the Collector's internal metrics through its own pipelines so they are enriched with resource attributes before being exported to Datadog.

## Further reading

{{< partial name="whats-next/whats-next.html" >}}

[1]: https://github.com/open-telemetry/opentelemetry-collector-contrib
[2]: /account_management/api-app-keys/
[3]: /getting_started/site/
[4]: /getting_started/tagging/unified_service_tagging/

> [5]: https://github.com/DataDog/opentelemetry-examples/tree/experimental-oss-config/configurations/opentelemetry-collector

Reviewer 4: Looks like this is pointing to a feature branch — should this instead point to `https://github.com/DataDog/opentelemetry-examples/tree/main/configurations/opentelemetry-collector`?

Author: @mx-psi - When do we expect to merge this into main?

Reviewer: I think that will have to wait until we're ready to GA the new recommended config. We can't really have a "private page" on a public repository like opentelemetry-examples, so I think it's best to keep it on a separate branch for the time being.

Reviewer 2: Yes, let's keep this as it is right now and revisit this when we make this page public
